src/Graph/GraphPrinter.php

```diff
@@ -87,197 +86,225 @@ class GraphPrinter extends SMWResultPrinter {
 
-		// Loop throught all fields of the record.
-		foreach ( $row as $i => $resultArray ) {
+				$nodeName = $node->getID();
 
-			// Loop throught all the parts of the field value.
-			while ( ( $object = $resultArray->getNextDataValue() ) !== false ) {
-				$propName = $resultArray->getPrintRequest()->getLabel();
-				$isName = $this->m_nameProperty ? ( $i != 0 && $this->m_nameProperty === $propName ) : $i == 0;
+				foreach ( $node->getParentNode() as $parentNode ) {
 
-				if ( $isName ) {
-					$name = $this->getWordWrappedText( $object->getShortText( $outputmode ), $this->m_wordWrapLimit );
-				}
+					// handle parent/child switch (parentRelation)
+					$graphInput .= $this->parentRelation ? " \"" . $parentNode['object'] . "\" -> \"" . $nodeName . "\""
+						: " \"" . $nodeName . "\" -> \"" . $parentNode['object'] . "\" ";
+
+					if ( $this->showGraphLabel || $this->showGraphColor ) {
+						$graphInput .= ' [';
+
+						// add legend item only if missing
+						if ( array_search( $parentNode['predicate'], $this->legendItem, true ) === false ) {
+							$this->legendItem[] = $parentNode['predicate'];
+						}
+
+						// assign color
+						$color = $this->graphColors[array_search( $parentNode['predicate'], $this->legendItem, true )];
 
-				if ( !( $this->m_nameProperty && $i == 0 ) ) {
-					$segments[] = $this->getGVForDataValue( $object, $outputmode, $isName, $name, $propName );
+						// show arrow label (graphLabel is misleading but kept for compatibility reasons)
+						if ( $this->showGraphLabel ) {
+							$graphInput .= "label=\"" . $parentNode['predicate'] . "\"";
+							if ( $this->showGraphColor ) {
+								$graphInput .= ",fontcolor=$color,";
+							}
+						}
+
+						// colorize arrow
+						if ( $this->showGraphColor ) {
+							$graphInput .= "color=$color";
+						}
+						$graphInput .= ']';
+					}
 				}
+				$graphInput .= ';';
 			}
 		}
+		$graphInput .= "}";
+
+		// Calls graphvizParserHook function from MediaWiki GraphViz extension
+		$result = $GLOBALS['wgParser']->recursiveTagParse( "<graphviz>$graphInput</graphviz>" );
 
-		return implode( "\n", $segments );
+		// append legend
+		$result .= $this->getGraphLegend();
+
+		return $result;
 	}
 
 	/**
-	 * Returns the GV for a single SMWDataValue.
+	 * Creates the graph legend
 	 *
-	 * @since 1.5.4
+	 * @since 3.1
 	 *
-	 * @param SMWDataValue $object
-	 * @param $outputmode
-	 * @param boolean $isName Is this the name that should be used for the node?
-	 * @param string $name
-	 * @param string $labelName
+	 * @return string Html::rawElement
 	 *
-	 * @return string
 	 */
-	protected function getGVForDataValue( SMWDataValue $object, $outputmode, $isName, $name, $labelName ) {
-		$graphInput = '';
-		$nodeLabel = '';
-		$text = $object->getShortText( $outputmode );
-
-		if ( $this->m_graphLink ) {
-			$nodeLinkURL = "[[" . $text . "]]";
-		}
+	protected function getGraphLegend(){
+		if ( $this->showGraphLegend && $this->showGraphColor ) {
```

Comment:
There is no return value outside of this block. Going to lead to errors in strict mode I figure. You can have a guard clause at the start here that returns an empty string.